### DISCUSSION TITLE
Decentralized multiroom audio solution

### DISCUSSION THREAD

**Discussion by rostom132** [snapcast/snapcast#1025] [Q&A]
How can I design a system with multiple devices that allows them to discover themselves automatically on the local network and synced together? Can a device become a server or client based on context? 

For example, if I connect to any arbitrary device in the system and input the audio stream, it can broadcast the audio to other devices.

**Comment by Piocky** [comment]
Hi, we do something quite similar right now using Zeroconf, when a device starts it looks for the Snapcast service on local network, if it founds no one it becomes the master (it starts snapserver), if it founds someone else it starts only snapclient. You can do some management using IP also if you want, like the bigger/smaller is always the master if you wish.

**Reply by artmg** [reply]
Using the [Android client](https://github.com/badaix/snapcast#android-client) will give you an idea of the way that snapcast already uses mDNS (aka Zeroconf / Bonjour) to make servers and clients visible, and to form groupings between them and control them. I came across a project yesterday, that appears to be focussed on having a constellation of devices across rooms or locations that could input/output/both as needed: https://github.com/Daenou/ansible-multiroom-audio - maybe you can get more ideas there?
